govcd/edgegateway_test.go

```diff
@@ -538,3 +542,111 @@ func (vcd *TestVCD) Test_UpdateNATRule(check *C) {
 }
+
+// TestEdgeGateway_UpdateLoadBalancerGlobal main point is to test that no load balancer configuration
+// xml tags are lost during changes of load balancer main settings (enable, logging)
+// The test does following steps:
+// 1. Cache raw XML body and marshaled struct in variables before running the test
+// 2. Toggle the settings of load balancer in various ways and ensure no err is returned
+// 3. Set the settings back as they originally were and again get raw XML body and marshaled struct
+// 4. Compare the XML text and structs before configuration and after configuration - they should be
+// identical except <version></version> tag which is versioning the configuration
+func (vcd *TestVCD) TestEdgeGateway_UpdateLoadBalancerGlobal(check *C) {
+	if vcd.config.VCD.EdgeGateway == "" {
+		check.Skip("Skipping test because no edge gatway given")
+	}
+	edge, err := vcd.vdc.FindEdgeGateway(vcd.config.VCD.EdgeGateway)
+	check.Assert(err, IsNil)
+
+	if !edge.HasAdvancedNetworking() {
+		check.Skip("Skipping test because the edge gateway does not have advanced networking enabled")
+	}
+
+	// Cache current load balancer settings for change validation in the end
+	beforeLb, beforeLbXml := testCacheLoadBalancer(edge, check)
+
+	lbModify, err := edge.GetLoadBalancer()
+	check.Assert(err, IsNil)
+
+	lbModify.Enabled = true
+	lbModify.AccelerationEnabled = true
+	lbModify.Logging.Enable = true
+	lbModify.Logging.LogLevel = "critical"
+
+	_, err = edge.UpdateLoadBalancerGlobal(lbModify)
+	check.Assert(err, IsNil)
+
+	lbModify.Enabled = false
+	lbModify.AccelerationEnabled = false
+	lbModify.Logging.Enable = true
+	lbModify.Logging.LogLevel = "emergency"
+	_, err = edge.UpdateLoadBalancerGlobal(lbModify)
+	check.Assert(err, IsNil)
+
+	// Try to set invalid loglevel to get validation error
+	lbModify.Logging.LogLevel = "invalid_loglevel"
+	_, err = edge.UpdateLoadBalancerGlobal(lbModify)
+	check.Assert(err, ErrorMatches, ".*Valid log levels are.*")
+
+	// Restore to initial settings and validate that it
+	_, err = edge.UpdateLoadBalancerGlobal(beforeLb)
+	check.Assert(err, IsNil)
+
+	// Validate load balancer configuration against initially cached version
+	testCheckLoadBalancerConfig(beforeLb, beforeLbXml, edge, check)
+}
+
+// testGetLoadBalancerXML is used for additional validation that modifying load balancer
+// does not change any single field. It returns a string of whole load balancer configuration
+func testGetLoadBalancerXML(edge EdgeGateway, check *C) string {
+
+	httpPath, err := edge.buildProxiedEdgeEndpointURL(types.LBConfigPath)
+	check.Assert(err, IsNil)
+
+	reqUrl, err := url.ParseRequestURI(httpPath)
+	check.Assert(err, IsNil)
+
+	req := edge.client.NewRequest(nil, http.MethodGet, *reqUrl, nil)
+
+	resp, err := checkResp(edge.client.Http.Do(req))
+	check.Assert(err, IsNil)
+
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
```

Comment:
Since we are bypassing `decodeBody`, could you add a call to `util.ProcessResponseOutput` here?
Or even better, could you make this operation an API call that avoids unmarshaling? We may need it for similar operations with other features.

Reply:
In fact the `ExecuteRequestWithCustomError` does it. There was a problem with reading `resp.Body` twice, but I think it is solved now.